crawl-ref/source/dgn-overview.cc

```diff
@@ -1093,3 +1062,164 @@ bool connected_branch_can_exist(branch_type br)
 
     return true;
 }
+
+/**
+ * Make the little overview
+ * (D) Dungeon        (T) Temple         (L) Lair           etc.
+ * at most 4 branches on 1 line
+*/
+static void _show_dungeon_overview(vector<branch_type> brs)
+{
+    clear_messages();
+    int linec = 0;
+    string line;
+    for (branch_type br : brs)
+    {
+        if (linec == 4)
+        {
+            linec = 0;
+            mpr(line);
+            line = "";
+        }
+        line += make_stringf("(%c) %-14s ",
+                             branches[br].travel_shortcut,
+                             branches[br].shortname);
+        ++linec;
+    }
+    if (!line.empty())
+        mpr(line);
+    return;
+}
+
+static int _prompt_annotate_branch(level_id lid)
+{
+    // these 3 lines make a vector containing all shown branches.
+    vector<branch_type> brs;
+    for (branch_iterator it; it; ++it)
+        if (is_known_branch_id(it->id))
+            brs.push_back(it->id);
+
+    mprf(MSGCH_PROMPT, "Annotate which branch? (. - %s, ? - help, ! - show branch list)",
+        lid.describe(false, true).c_str());
+
+    while (true)
+    {
+        int keyin = get_ch();
+        switch (keyin)
+        {
+        CASE_ESCAPE
+            return ID_CANCEL;
+        case '?':
+            show_annotate_help();
+            break;
+        case '!':
+            _show_dungeon_overview(brs);
+            break;
+        case '\n': case '\r': case '.':
+            return ID_HERE;
+        case '<':
+            return ID_UP;
+        case '>':
+            return ID_DOWN;
+        case CONTROL('P'):
+            {
+                const branch_type parent = parent_branch(lid.branch);
+                if (parent < NUM_BRANCHES)
+                    return parent;
+            }
+            break;
+        default:
+            // Is this a branch hotkey?
+            for (branch_type br : brs)
+            {
+                if (toupper_safe(keyin) == branches[br].travel_shortcut)
+                    return br;
+            }
+            // Otherwise cancel
+            return ID_CANCEL;
+        }
+    }
+}
+
+void do_annotate()
+{
+    const level_id lid  = level_id::current();
+    const int branch = _prompt_annotate_branch(lid);
+
```

Comment:
Given the way int is used here, I might recommend the following paradigm rather than the if sequences:

```
if (branch < 0)
{
    ASSERT(branch >= FIRST_ANNOTATION_POSSIBILITY && branch <= LAST_ANNOTATION_POSSIBILITY); // you would need to add these to the enum (and probably name them differently if you adopt my enum rename suggestion in a different comment)
    AnnotationPossibilities a = static_cast<AnnotationPossibilities>(branch);
    switch(a)
    {
    case ID_CANCEL:
        ...
    }
}
else
{
    ...
}
```

The reason is that the switch will check that every enum value is used so it provides a compile-time guard in case someone later changes the enum.

It might also be worth adding some ASSERTions to validate the positive case, i.e. at least that `branch < NUM_BRANCHES`

If this were still to be done with an `if` sequence, I would still suggest using `if ... else if` rather than unconnected `if`s, and adding the validation checks.